Сalendar module with the ability to add, delete, edit a named event and receive events for a certain period

AlehBarushka opened the pull request.

AlehBarushka committed:
added module Calendar with methods(addEvent, deleteEvent)

AlehBarushka committed:
testing the module

AlehBarushka commented:
need to add documentation and an html page for testing from the console

AlehBarushka committed:
deleted index.js

AlehBarushka committed:
install jsdoc

AlehBarushka committed:
updated methods and added comments

AlehBarushka committed:
added index.html with ref on documentation

AlehBarushka committed:
added jsdoc

vitali-working commented on module/Calendar.js:
```diff
@@ -0,0 +1,64 @@
+	 * addEvent({id: 1, title: 'birthday',startTime: '2022-04-06T12:00:00',}, () => { console.log('Happy Birthday') })
+	 *
+	 */
+	addEvent(event, callback) {
```
What's happened if event / callback will be empty?

vitali-working commented on module/Calendar.js:
```diff
@@ -0,0 +1,64 @@
+		if (index !== -1) {
+			this.events.splice(index, 1);
+		}
+		clearTimeout(event.timerId);
```
if user pass empty event?

AlehBarushka replied:
If event is empty, there will be an error

AlehBarushka committed:
Added error handling for missing parameters in methods

AlehBarushka committed:
added event object checking for empty fields

AlehBarushka commented:
Added error handling of addEvent and deleteEvent methods:
- method parameters are required;
- the event parameter object must contain the id, startTime fields for the addEvent method;
- the event parameter object must contain the id, timerId fields for the deleteEvent method.

newqwes commented on package.json:
```diff
@@ -15,6 +16,8 @@
-	"homepage": "https://github.com/AlehBarushka/CalendarOfEvents#readme"
+	"homepage": "https://github.com/AlehBarushka/CalendarOfEvents#readme",
+	"devDependencies": {
+		"jsdoc": "^3.6.10"
```
It seems to me that this was not meant, but simply use the description of the methods through jsDoc (without the jsDoc library). Extra files are created in the 'out' folder, which distracts from the main task. WDYT @vitali-bantsevich ?

AlehBarushka replied:
Yes, at first I misunderstood a little) 
deleted JSDoc from dev dependencies

AlehBarushka committed:
changed the Calendar object to class, added getAllEvents and SetEvent methods, changed the events field to private

AlehBarushka committed:
deleted the generated JSDoc package files

AlehBarushka committed:
deleted sample

AlehBarushka committed:
added private methods for validate, changed the addEvent and deleteEvent methods, added a private method for generating idaadded priv

AlehBarushka committed:
deleted _setEvent, updated _eventValidator, updated _dateValidator,  updated addEvent, updated deletedEvent, added updateEvent,

addEvent: added callback in event obj, added return status;
deleteEvent: id parameter instead of event obj, added return status;
updateEvent: added new method.

AlehBarushka committed:
added getEventsForPeriod method

AlehBarushka committed:
updated getEventsForPeriod, added getEventForDay

AlehBarushka committed:
updated getEventsForPeriod, updated getEventForDay, added getEventForMonth method

newqwes commented on module/Calendar.js:
```diff
@@ -0,0 +1,311 @@
+	 * random string
+	 */
+	_getId() {
+		const s4 = () => {
```
_s4_ - wrong naming

AlehBarushka replied:
renamed to generateRandomString

newqwes commented on module/Calendar.js:
```diff
@@ -0,0 +1,311 @@
+	 * @returns {string}
+	 * random string
+	 */
+	_getId() {
```
This method generates a random string(what the _string_  maybe uniqID?), and the name of the method says that we get the id. Problem with naming

AlehBarushka replied:
fixed to "The method generates and returns a uniqId"

newqwes commented on module/Calendar.js:
```diff
@@ -0,0 +1,311 @@
+			const validatedDate = this._dateValidator(event.date, event.time);
+			const timerId = setTimeout(
+				callback,
+				new Date(validatedDate).getTime() - Date.now()
```
It would be more understandable if you put it in a separate variable, as you wish.
`const delay = new Date(validatedDate).getTime() - Date.now()`

AlehBarushka replied:
Thank you. I fixed it. 
I like your version better)

newqwes commented on module/Calendar.js:
```diff
@@ -0,0 +1,311 @@
+			);
+			const eventObj = {
+				id: this._getId(),
+				...validatedEvent,
```
`validatedEvent` can accidentally overwrite your `id`, you can put it in up of `id`, or add the necessary destructuring to the `eventValidator` method.

AlehBarushka replied:
Fixed. Made the destructurization of the object

newqwes commented on module/Calendar.js:
```diff
@@ -0,0 +1,311 @@
+class Calendar {
```
For better readability, please use a tab consisting of two spaces

newqwes replied:
Also use _eslint_ to check the code

AlehBarushka replied:
OK, now i'm using two spaces.

newqwes commented on module/Calendar.js:
```diff
@@ -0,0 +1,311 @@
+		}
+	}
+
+	/**
```
Use this _jsDoc_ method description approach everywhere (excessive description for parameters)

AlehBarushka replied:
Does it mean not to use "@method" in the description?

newqwes replied:
> Does it mean not to use "@method" in the description?

Yes, this is a redundant action, we already see the name of the method, it’s not worth clogging up with extra lines of code.

AlehBarushka replied:
ОК! Fixed

newqwes commented on module/Calendar.js:
```diff
@@ -0,0 +1,311 @@
+
+	/**
+	 * The method returns array with events objects in the specified date range
+	 * @method getEventsForPeriod
```
`@method getEventsForPeriod` - redundant description, also in other places.

AlehBarushka replied:
Fixed

newqwes commented on module/Calendar.js:
```diff
@@ -0,0 +1,311 @@
+			const parsedDateRange2 = Date.parse(dateRange2);
+			const events = this._events.filter((event) => {
+				return (
+					Date.parse(event.date) >= parsedDateRange1 &&
```
Move to variable

AlehBarushka replied:
moved to variable

newqwes commented on module/Calendar.js:
```diff
@@ -0,0 +1,311 @@
+			const parsedDateRange2 = Date.parse(dateRange2);
+			const events = this._events.filter((event) => {
+				return (
+					(Date.parse(event.date) >= parsedDateRange1 &&
```
Move to variable

AlehBarushka replied:
moved to variable

AlehBarushka replied:
moved to variable

newqwes commented on module/Calendar.js:
```diff
@@ -0,0 +1,311 @@
+			if (isNaN(Date.parse(date))) {
+				throw new Error('Incorrect date format!');
+			}
+			const parsedDate = Date.parse(date);
```
Duplicating code, calling `Date.parse(date)` twice. Take it to a variable

AlehBarushka replied:
fixed

AlehBarushka replied:
moved to variable

newqwes commented on module/Calendar.js:
```diff
@@ -0,0 +1,311 @@
+				return Date.parse(event.date) === parsedDate;
+			});
+			if (events.length) {
+				return events[0];
```
Why do we `return events[0];` only one event, what if there will be more than one event on that day?

AlehBarushka replied:
fixed

AlehBarushka replied:
fixed

newqwes commented on module/Calendar.js:
```diff
@@ -0,0 +1,311 @@
+				throw new Error('Incorrect date format!');
+			}
+			const nextMonth = (month) => {
+				if (parseInt(month) + 1 < 10) {
```
`parseInt(month) + 1` - code duplication, put into a variable.

AlehBarushka replied:
moved to variable

AlehBarushka replied:
moved to variable

newqwes commented on module/Calendar.js:
```diff
@@ -0,0 +1,311 @@
+			if (isNaN(Date.parse(dateRange1))) {
+				throw new Error('Incorrect date format!');
+			}
+			const nextMonth = (month) => {
```
The code _smells_, a method declaration inside a method. There is also a large nesting. Same throughout the code.

vitali-working commented on module/Calendar.js:
```diff
@@ -0,0 +1,311 @@
+	 * The method checks the validity of the event object to use it in the addEvent method
+	 * @private
+	 * @method _addEventValidator
+	 * @param {object} event
```
Please, define type for _event_

AlehBarushka committed:
deleted jsDoc from dev dependencies

AlehBarushka committed:
fixed naming of the _generateUniqId method and description

AlehBarushka committed:
added delay

AlehBarushka committed:
fixed _eventValidator, addEvent and updateEvent methods

_eventValidator - added title checking;
addEvent, updateEvent  - made the destructurization of the event object, updated error handling.

AlehBarushka committed:
formatted the code using a tab consisting of two spaces

AlehBarushka committed:
fixed some error handling, duplicates moved to variable, fixed getEventsForDay return

AlehBarushka committed:
deleted redundant description of methods

AlehBarushka committed:
updated jsDoc descriptions

AlehBarushka committed:
Merge branch 'feature' into feature-test

AlehBarushka committed:
some update

newqwes commented on module/Calendar.js:
```diff
@@ -0,0 +1,321 @@
+   * @description The method generates and returns a uniqId.
+   * @returns {string} unique Id
+   **/
+  _generateUniqId() {
```
I think this method does not belong to the calendar class, it can be moved to utilities.

AlehBarushka committed:
generateUniqId moved to utilities
